osquery/dispatcher/tests/task_schedule_tests.cpp

```diff
@@ -0,0 +1,145 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under both the Apache 2.0 license (found in the
+ *  LICENSE file in the root directory of this source tree) and the GPLv2 (found
+ *  in the COPYING file in the root directory of this source tree).
+ *  You may select, at your option, one of the above-listed licenses.
+ */
+
+#include <chrono>
+#include <iostream>
+#include <thread>
+
+#include <gtest/gtest.h>
+
+#include <osquery/config.h>
+#include <osquery/logger.h>
+#include <osquery/system.h>
+
+#include "osquery/dispatcher/task_schedule.h"
+
+namespace osquery {
+
+DECLARE_bool(disable_logging);
+
+class TaskScheduleTests : public testing::Test {
+  void SetUp() override {
+    logging_ = FLAGS_disable_logging;
+    FLAGS_disable_logging = true;
+  }
+
+  void TearDown() override {
+    FLAGS_disable_logging = logging_;
+  }
+
+ private:
+  bool logging_{false};
+};
+
+TEST_F(TaskScheduleTests, run_for_the_one_time) {
+  auto tasks = TaskSchedule{};
+  auto counter = int{0};
+  tasks.add(
+      [&counter](const auto scheduled_time) {
+        ++counter;
+        return 0;
+      },
+      0);
+  while (!tasks.isEmpty()) {
+    auto next_time_to_run = tasks.nextTimeToRun();
+    auto now = getUnixTime();
+    if (now < next_time_to_run) {
+      std::this_thread::sleep_for(std::chrono::seconds{next_time_to_run - now});
+    } else {
+      tasks.runNextNow();
+    }
+  }
+  EXPECT_EQ(counter, 1);
+}
+
+TEST_F(TaskScheduleTests, first_start_priority) {
+  auto tasks = TaskSchedule{};
+  auto counter = int{0};
+  tasks.add(
+      [&counter](const auto scheduled_time) {
+        ++counter;
+        return 0;
+      },
+      10);
+  tasks.add(
+      [&counter](const auto scheduled_time) {
+        counter *= 2;
+        return 0;
+      },
+      100);
+  while (!tasks.isEmpty()) {
```

Comment:
Again no need to sleep. Just ensure all values are returned according to what you expect, e.g. `nextTimeToRun()` returns `10` then after `runNextNow()` `100`.